### ISSUE TITLE
What's going on with the lack of new exchange additions?

### ISSUE THREAD

**Issue by hanyoonLDA** [ccxt/ccxt#5822]
It seems that @kroitor is the only one that is responsible for helping with exchange integrations, reviewing the code, and merging it with the main branch. It definitely does not sound like a fun job. Is there any way the community can help in terms of workload and due diligence? The entire community should be grateful not only for this project, but to @kroitor for being the seemingly only active dev that is still helping with the "boring" parts. 

There hasn't been any new exchanges in a while, and it seems like a crapshoot in estimating the time to be approved. It's gotten so bad that I've been adding new exchanges manually for the ones that passed all the automated checks. 

Please don't hesitate to ask the community for help -- after all, that's what open source is all about right? :-) The unified API and strict contribution guidelines may be an annoyance to some, but I think it's the best thing since sliced bread. However, there is no denying that CCXT brings a lot of volume to new exchanges, and the waiting time can get them a bit anxious 😅 And to everyone using CCXT, make some donations when you have some extra Sats or Ether laying around. _Especially_ if you're profiting from it.

**All hail the greatest open source crypto trading platform in existence.** All hail @kroitor... but someone please help him out 😂

**Comment by kroitor** [comment]
> There hasn't been any new exchanges in a while, and it seems like a crapshoot in estimating the time to be approved. It's gotten so bad that I've been adding new exchanges manually for the ones that passed all the automated checks.

Yes, that's a known issue. In fact, this entire matter is so complicated, that it's very hard to cover in one post ) In general the CCXT has gained a lot of traction for the speed with which new features, functions and exchanges were added. Having many/new exchanges is one of CCXT's selling points. 

As you've noticed precisely, starting this summer the activity on new integrations slowed down a bit. We are still integrating at least a few exchanges every month, but, obviously, there's more of them than we are adding now. We've somewhat changed our working time spent on new integrations in favor of the work on adding the WS support that is going to be released in CCXT Pro in a few weeks, we believe ([more about it here](https://github.com/ccxt/ccxt/issues/56#issuecomment-507290536)). 

After that we are going to merge pending PRs in bulk and will restore the former pace of at least several new integrations each week. We plan for adding several dozens or up to a hundred of new exchanges this fall.

> Please don't hesitate to ask the community for help -- after all, that's what open source is all about right? :-) 

The main reason why pending PRs are still pending is the lack of code quality.

To be honest, we are very opinionated when it comes to merging, and this is my bad, I know. I'm idealistic and pedantic on new additions, despite that the lib might already have inside it many things that aren't ideally implemented (not even close). I believe that helps protect the lib from being polluted. I don't hesitate to ask the community for help, in fact, I do ask for help openly: dear community, please help! ) But it's a kind of a chicken and an egg problem if you think of it. Because those PRs come from the community in the first place. And they lack code quality. 

Again, I'm not saying, that the lib currently has high quality code everywhere, there's tons of bugs living there. And it may sound like ungrateful of me to say that the code that comes from the community is not good enough. The true reasoning behind not merging the PRs is to maintain the unified standard at least to some extent and to ensure that the lib does not break down in simple cases – so that it works for the users, that is, for the community.

Now, if those PRs come from the community, asking the community to work more on those same PRs doesn't help, since the community cannot (or very rarely can) produce the code quality that is required for it to work seamlessly across all languages in a unified manner (which is one of the main reasons for you to use this lib in the first place – if it didn't work you would not use it). 

The errors in the existing PRs can be classified in several categories:

1. kindergarden errors (due to a lack of programming skills)
2. violations of contributing guidelines (despite that it's written that they have to be followed literally to a word, people just don't rtfm), including repetetive reinvention of same old bugs over and over again instead of copying from existing exchanges
3. complex things and really new development problems that have not been encountered yet

(1) and (2) make up most of the errors, 90% of them. That means 90% of the code coming from the community is very far from ideal (I'm sorry, because I really love our community, and I don't want to screw my karma at all). 

Very few contributors actually submit quality edits, and it usually doesn't go right at first attempt anyway. So, it takes some time to play with the lib, get familiar with the structure and also get to a level that allows consistently producing code that will work and won't break down on stupid mistakes. People who do submit quality code usually get their PRs merged within minutes, if not seconds.  Topmost contributors already know how to do that, so I don't even have to ask them (infinite thanks).

Unfortunately, I can't point out each and every single error, because there's so many of them and they are duplicates of things, that have been covered, explained and solved already. I can't teach programming en mass so that we suddenly start getting code without obvious beginner-level errors, like not returning values where expected, missing type checks, not reusing the existing base code, not sanitizing input, improper work with symbols and ids, missing fields, redundant fields, breaking the unification, etc, etc... It just takes less time to go and fix the error myself, than to explain why it's an error and why it should be fixed (mind that the explanation is outlined in the contributing guidelines, but people just don't read). And I can't tell the author _"sorry, there's some bad code there and I will have to check it line by line, or, in other words, do a complete rework/rewrite of it"_. Most of those errors are logical errors, so, we can't automate the process of spotting those errors, it very often requires human+eyes+attention+brain+experience. 

Our testing framework isn't ideal either – if it was good enough, we would be able to catch many of the errors automatically. Enhancing the testing framework is actually the way to partially solve the problem and help speed up the merging process, so that we could merge new submissions faster and faster.

Thus, we're in a situation where the community submits a PR and then I can't ask the same community to help me clean that PR or speed things up, because this is kinda already implied and what we have is what we have – that's the max quality that we will get. If the community could see and self-correct own mistakes in the code of its own PRs, those PRs would have been merged long ago. And if some of them aren't merged, that means, the community _does not see_ the errors in the code, and then those errors have to be found and seen first, and then corrected second. I can't really merge or comment on them in a detailed manner while I'm focused on streaming. Therefore those PRs have to wait a few weeks till I get to them. Unless someone polishes them first. Which is unlikely, because people are awaiting for a review from me. Thus, a chicken and an egg problem. But I believe most of those PRs will be merged this autumn among many other new exchanges. And we also hope that the CCXT Pro release will help us fund the work on new integrations, to invite more devs and expand the team.

> However, there is no denying that CCXT brings a lot of volume to new exchanges, and the waiting time can get them a bit anxious 😅 

Totally agree with you on this! Thx! )

Let me know if the above does not answer it.

**Comment by frosty00** [comment]
https://i.imgur.com/db5Myu4.png

**Comment by kroitor** [comment]
Just listed Binance US. 

Also: https://twitter.com/ccxt_official

Closing this for now, feel free to open it if needed or just ask further questions, if any )

**Comment by brandsimon** [comment]
@kroitor 

> Again, I'm not saying, that the lib currently has high quality code everywhere, there's tons of bugs living there.

Does it make sense to document an exchange, which is at the current state of the lib? So new developers dont use an old/bad/buggy exchange as template for new exchanges.

> The true reasoning behind not merging the PRs is to maintain the unified standard at least to some extent and to ensure that the lib does not break down in simple cases

Do you think it would help new developers to add some methods that check the return value of unified return values?

**Comment by kroitor** [comment]
@brandsimon 

> Does it make sense to document an exchange, which is at the current state of the lib? So new developers dont use an old/bad/buggy exchange as template for new exchanges.

Yep, that's a good idea, thanks! Obviously, this is something I will do asap. Will list out the exchanges that are "good" in the contributing guidelines.

> Do you think it would help new developers to add some methods that check the return value of unified return values?

Well, the point of running the tests on unified structures is to do just that – check the return values, however, the problem is – we don't have all the checks that would guarantee non-breaking code. Unfortunately, our test set does not filter out all bugs, and it requires spending actual money for private API tests (withdrawals, orders, fills, fees, etc). The problem with automating those private tests boils down to privacy and a setup with encrypted API keys on travis that would be abuse-resistant...

**Comment by kroitor** [comment]
@brandsimon 

- https://github.com/ccxt/ccxt/wiki/Requirements
- https://github.com/ccxt/ccxt/blob/master/js/test/Exchange/test.currency.js
- https://github.com/ccxt/ccxt/blob/master/js/test/Exchange/test.market.js
- https://github.com/ccxt/ccxt/blob/master/js/test/Exchange/test.ohlcv.js
- https://github.com/ccxt/ccxt/blob/master/js/test/Exchange/test.order.js
- https://github.com/ccxt/ccxt/blob/master/js/test/Exchange/test.orderbook.js
- https://github.com/ccxt/ccxt/blob/master/js/test/Exchange/test.ticker.js
- https://github.com/ccxt/ccxt/blob/master/js/test/Exchange/test.trade.js
- https://github.com/ccxt/ccxt/blob/master/js/test/Exchange/test.transaction.js
- https://github.com/ccxt/ccxt/tree/master/js/test/Exchange

**Comment by brandsimon** [comment]
@kroitor 
Great, thank you very much for the documentation.

Sorry, I should have been more clear, I dont mean automated tests, I meant like manual test functions.
When I add a new exchange, I need to do testing anyway (creating / canceling orders), so I think on my side I could run
```
order = exchange.fetch_order(id)
print_unified_order_structure_errors(order)
```

Another thought was parsing all return values, so in the exchange you would have to do
```
async fetchBalance (params = {}) {
    // Do stuff
    return return_and_validate_unified_order_structure(order);
}
```
I dont think this is ideal. It adds some overhead. The function should never throw, just log errors.

**Comment by kroitor** [comment]
@brandsimon here's how you would currently run the tests on those unified structures at present time:

1. create `keys.local.json` containing `{"exchange_id":{"apiKey":"...", "secret": ..."}}`
2. make sure the list of exchange ids is built `npm run export-exchanges`
3. run `node js/test/test exchange_id` → if your credentials are correct, it should test the private part as well

↑ The problem is: those tests are incomplete, and have to be extended to cover all cases and all possible errors.

**Comment by hanyoonLDA** [comment]
Thanks for the super detailed explanation @kroitor ! Again, I'm not trying to talk down the project or anything, I think you guys are doing gods work. Keep on at it! 

With respect and love,
LDA

**Comment by kroitor** [comment]
@brandsimon got back to this btw:

> Does it make sense to document an exchange, which is at the current state of the lib?

There is no single exchange that demonstrates all the tricks, so it has to be a list of several exchanges, but I will get it done asap. Apart from that I've actually started the work on the `skeleton.js` file which would contain a reference implementation of all possible methods. A developer would just copy and edit it in a few places in order to add an integration (with simple edits, like, cutting out excessive methods, leaving the applicable ones in place). I got a bit sidetracked in the process, so the progress on it has stuck and it's too raw to upload atm. But, I hope, a combination of a list of exchanges for reference + a skeleton.js template would work much better than it works now.
